### ISSUE TITLE
Support a growing window

### ISSUE THREAD

**Issue by bqback** [cogentcore/core#1298]
### Describe the bug

Providing GrowWrap = true to a new body makes it unable to display any content. Settings are taken directly from https://www.cogentcore.org/core/advanced/styling/.

Expected behavior: the windows contains a text widget with "Picker" and a picker with options "foo" and "bar". The window wraps tightly around contents of the child frame.

Result: a black window appears with a 0,0 box 
![result](https://github.com/user-attachments/assets/a6bae8a8-4bb4-48cd-9e44-78a78a4efb37)

### How to reproduce

`core run` the code in the issue

### Example code

```go
package main

import (
	"fmt"

	"cogentcore.org/core/core"
	"cogentcore.org/core/events"
	"cogentcore.org/core/styles"
	"cogentcore.org/core/styles/units"
)

func main() {
	b := core.NewBody()
	b.Styler(func(s *styles.Style) {
		s.GrowWrap = true
		s.Gap.X = units.Value{Value: 0.5, Unit: units.UnitEm}
		s.Gap.Y = units.Value{Value: 0.5, Unit: units.UnitEm}
	})

	fr := core.NewFrame(b)
	fr.Styler(func(s *styles.Style) {
		s.Align = styles.AlignSet{Items: styles.Center}
	})

	t := core.NewText(fr).SetText("Picker")
	t.Styler(func(s *styles.Style) {
		s.SetTextWrap(false)
	})
	ch := core.NewChooser(fr).SetStrings("foo", "bar")
	ch.OnChange(func(e events.Event) {
		core.MessageSnackbar(fr, fmt.Sprintf("Selected %v", ch.CurrentItem.Value))
	})

	b.RunMainWindow()
}
```

### Platform

Windows

**Comment by bqback** [comment]
I've also tried 
* styling the body after adding the frame
* removing `s.SetTextWrap`

with no success.

**Comment by kkoreilly** [comment]
Thank you for reporting this. `GrowWrap` is only intended for text elements, so it does not work to set that on the `Body`. You can try setting `Wrap` and/or `Grow` separately depending on what you are trying to accomplish; I can help you with the appropriate styling options once you tell me your use case. Thank you!

**Comment by rcoreilly** [comment]
we should probably make sure `GrowWrap` is properly ignored where it is irrelevant.

**Comment by bqback** [comment]
> I can help you with the appropriate styling options once you tell me your use case.

The full usecase would be a sort of "growing" app window with the scenario being:
* User is only show a single picker with two options and a caption
* Based on the option selected one of two frames with different sets of inputs (text field and two time pickers/a file picker) appears
* Some processing happens for a short while, with a cancel button that would go back a step, likely with logs in a readonly text area
* End result is presented as a large scrollable spreadsheet with an export button in a separate window/page

I'd like for the window to be automatically resized to wrap the visible frames up to step 3. I can vaguely predict the size and resize manually when frames are enabled/disabled, but that seems quite ineffective.

**Comment by rcoreilly** [comment]
See #1303 for discussion of why we cannot easily prevent the mistaken use of GrowWrap.

It should be relatively straightforward to add a `ResizeToContents`  function that you can call whenever you want. Currently we have a "PrefSize" function (which we will rename) that is used when a new window is opened, but there is no reason why it cannot be called to get a size that should hold a given Scene's contents.

**Comment by kkoreilly** [comment]
Yes, so once we add the `ResizeToContents` function soon, you will be able to call it at each of those stages you mentioned to get the window to grow. Until then, you can just have the window at a constant size without any `GrowWrap`. We will let you know once we implement the `ResizeToContents` function.

**Comment by bqback** [comment]
Thank you! One more thing to look forward to in 0.4 (another is localized time picker formatting).

Last question that's only tangentially related (because I mentioned text fields somewhere in the thread) -- how exactly is text field's `validate()` supposed to work? I understand how the validator is assigned, and when and how it's ran, but it seems to me like no values are passed to `Validator()`. It's just a function that returns an error, but what is it supposed work with?

I think the only option that would make sense to me is explicitly defining my validator as a bound method `func (tf *TextField) Validator() error` to have access to `tf.text` inside the method, but that doesn't feel right.

**Comment by bqback** [comment]
Could it possibly have been intended as a `func(string) error` function, with `tf.validate()` calling it as `tf.Validator(tf.text)`?

**Comment by bqback** [comment]
> Could it possibly have been intended as a `func(string) error` function, with `tf.validate()` calling it as `tf.Validator(tf.text)`?

Nvm, I've taken a look at tests for the validator, and it has to be defined in-place as an anonymous function. Don't really have any other questions, then.

**Comment by kkoreilly** [comment]
Yes, it is intended to be defined as an anonymous closure that calls `tf.Text()` or any other functions to get relevant attributes. That structure allows it to be more generically extensible. If you want, you could always define another function that just takes the text field or text string directly and call that function in the anonymous closure.

**Comment by kkoreilly** [comment]
We implemented support for growing windows in #1306, so you can do this if you update to the latest version on the main branch of Cogent Core:

```sh
go get cogentcore.org/core@43c6bd6
```

To get this to work, this is the general approach:

```go
func main() {
	b := core.NewBody("Growing window")
	bt := core.NewButton(b).SetText("Expand")
	bt.OnClick(func(e events.Event) {
		core.NewTextField(b)
		b.Scene.ResizeToContent()
	})
	b.RunMainWindow()
}
```

The one issue with that is the initial size of the window. You can bypass that issue by disabling the default window min size and renaming the window before you resize it so the larger size doesn't get saved for the starting state:

```go
func main() {
	b := core.NewBody("Starting window")
	bt := core.NewButton(b).SetText("Expand")
	bt.OnClick(func(e events.Event) {
		core.NewTextField(b)
		b.SetTitle("Expanded window")
		b.Scene.ResizeToContent()
	})
	b.NewWindow().SetUseMinSize(false).Run()
	core.Wait()
}
```

Please let me know if that works for you. If you have any issues, I would be glad to help.
